### ISSUE TITLE
`comma-dangle` incorrectly requires commas after rest properties

### ISSUE THREAD

**Issue by daltones** [eslint/eslint#7297]
**What rule do you want to change?**
comma-dangle

**Does this change cause the rule to produce more or fewer warnings?**
I guess it depends. But probably more warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, something like `{forbidAfterRest: true}`

**Please provide some example code that this change will affect:**
```js
const {
  x,
  y,
  ...z
} = foo;
```

**What does the rule currently do for this code?**
When using this rule with string option `"always-multiline"`, it will warn about missing trailing comma after `...z`.

**What will the rule do after it's changed?**
No trailing comma after `...z` will be correct, and a trailing comma will be incorrect.

**Motivation:**
Seems like the [spec](http://sebmarkbage.github.io/ecmascript-rest-spread/) doesn't even allow that kind of trailing comma.

[Babel's Babylon v6.11.3](https://github.com/babel/babylon/releases/tag/v6.11.3) went so far as to consider that as syntax error.
> The rationale for this is that the use case for trailing comma is that you can add something at the end without affecting the line above. Since a RestProperty always has to be the last property it doesn't make sense.

[Babylon v6.11.4](https://github.com/babel/babylon/releases/tag/v6.11.4) temporary rolled back that behavior.

So for now I guess is up to the user to choose if that trailing comma is incorrect or not with the new `forbidAfterRest` option.

**Comment by not-an-aardvark** [comment]
If having a trailing comma after a spread property is a syntax error, then there would be no point in having a separate option for it, since a parsing error would get thrown before the option had a chance to take effect.

However, we should ensure that the rule doesn't warn the user about missing trailing commas if adding the comma would result in a SyntaxError.

It seems to already do this with array spread elements:

```js
/* eslint comma-dangle: [2, "always"] */

var [
  foo,
  ...bar // No error is reported (correct behavior)
] = baz;
```

**Comment by not-an-aardvark** [comment]
I was able to reproduce the issue.

Config:

```json
{
  "rules": {
    "comma-dangle": [2, "always"]
  },
  "parserOptions": {
    "ecmaVersion": 8,
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true
    }
  }
}
```

Code:

```js
var {foo, ...bar} = baz;
```

Expected no errors, but one error was reported:

```
1:17  error  Missing trailing comma  comma-dangle
```

This should not be reported as an error, since a trailing comma after a spread property is a SyntaxError according to the [spec](https://sebmarkbage.github.io/ecmascript-rest-spread/#Rest). (At the moment it seems like acorn parses that successfully, which is probably a separate bug.)

**Comment by daltones** [comment]
@not-an-aardvark yes, this rule change proposal could be rethought as a bug report to Espree. But I was considering the Babylon case, which rolled back that SyntaxError because reported compatibility problems, making the trailing comma optional.

**Comment by daltones** [comment]
@not-an-aardvark Understood. So we can conclude that for linting purposes there would be no option about the trailing comma after array or object rest property, it's a matter of syntax. And for now we just found a false-positive for comma-dangle.

**Comment by not-an-aardvark** [comment]
I suppose there are a few things going on here:

* At the moment, the spec forbids trailing commas following spread properties. We have no reason to believe that the spec will change in this regard before it's finalized, especially since commas are already forbidden after spread elements in arrays.
* Acorn and Babel have slightly incorrect parsing at the moment; they don't throw errors with trailing commas after spread elements.
* When engines implement spread properties, they will presumably throw a SyntaxError at runtime if they encounter a trailing comma after a spread property.


At the moment, the `comma-dangle` rule requires people to use trailing commas after spread properties, and it autofixes by adding commas. I think this is a significant problem, because peoples' code will be syntactically invalid according to the spec.

So I think the best thing to do would be to fix the false-positive for `comma-dangle`, as you mentioned.

**Comment by daltones** [comment]
@not-an-aardvark Sure, you're right. Thanks!

**Comment by hzoo** [comment]
some more context: It was pushed out as a patch version in babylon but it "broke" react-native. It was failing on react-native's own code so the user's would have to either change node_modules or wait for react-native to update. Thus we reverted it.

**Comment by ljharb** [comment]
… just to clarify, trailing commas are always permitted after spreads, they're just never permitted after rests - in both iterable or object destructuring, as well as in function argument signatures.

**Comment by not-an-aardvark** [comment]
Whoops, looks like I was a bit careless about that distinction when renaming the issue. (I think the fix does distinguish between rest and spread correctly, though.)
